wakuv2/waku.go

```
@@ -317,51 +303,96 @@ func (w *Waku) dnsDiscover(enrtreeAddress string, protocol libp2pproto.ID, apply
-}
+	identifyWg := &sync.WaitGroup{}
+	identifyWg.Add(len(cfg.WakuNodes))
+	for _, addrString := range cfg.WakuNodes {
+		if strings.HasPrefix(addrString, "enrtree://") {
+			// Use DNS Discovery
+			go func() {
+				w.dnsDiscover(ctx, idService, addrString)
+				identifyWg.Done()
+			}()
+		} else {
+			// It is a normal multiaddress
+			addr, err := multiaddr.NewMultiaddr(addrString)
+			if err != nil {
+				log.Warn("invalid peer multiaddress", addrString, err)
+				continue
+			}
 
-func (w *Waku) addWakuV2Peers(cfg *Config) {
-	if !cfg.LightClient {
-		addRelayPeer := func(m multiaddr.Multiaddr, protocol libp2pproto.ID) {
-			go func(node multiaddr.Multiaddr) {
-				ctx, cancel := context.WithTimeout(context.Background(), requestTimeout)
-				defer cancel()
-				err := w.node.DialPeerWithMultiAddress(ctx, node)
-				if err != nil {
-					log.Warn("could not dial peer", err)
-				} else {
-					log.Info("relay peer dialed successfully", node)
-				}
-			}(m)
+			go func(ma multiaddr.Multiaddr) {
+				identifyAndConnect(ctx, cfg.LightClient, idService, ma)
+				identifyWg.Done()
+			}(addr)
 		}
-		w.addPeers(cfg.RelayNodes, relay.WakuRelayID_v200, addRelayPeer)
 	}
 
-	addToStore := func(m multiaddr.Multiaddr, protocol libp2pproto.ID) {
-		peerID, err := w.node.AddPeer(m, protocol)
+	identifyWg.Wait()
+	return nil
+}
+
+func identifyAndConnect(ctx context.Context, isLightClient bool, identifyService *identify.IDService, ma multiaddr.Multiaddr) {
+	peerInfo, err := peer.AddrInfoFromP2pAddr(ma)
+	if err != nil {
+		log.Error("could not extract peerinfo", zap.Any("ma", ma), zap.Error(err))
+		return
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, 3*time.Second)
+	defer cancel()
+
+	err = identifyService.Host.Connect(ctx, *peerInfo)
+	if err != nil {
+		log.Error("could not extract peerinfo", zap.Any("ma", ma), zap.Error(err))
+		return
+	}
+
+	conns := identifyService.Host.Network().ConnsToPeer(peerInfo.ID)
+	if len(conns) == 0 {
+		return // No connection
```

shall we return an error in that case?

---

hm, maybe logging an error but returning an error i think it's not necessary. I'm not sure how would I handle a case where the connection to a peer was succesful, yet libp2p Network() says that there's no connection to that peer